src/interval.jl

```diff
@@ -351,6 +363,266 @@ true
+track(x::Endpoint, ::TrackEachEndpoint) = isclosed(x)
+track(x::Bool, ::TrackEachEndpoint) = x
+# but only if required
+track(_, ::TrackStatically) = nothing
+# combine tracked endpoints into a final descion for a new endpoint
+# using the appropriate boolean operator (e.g. || for union, && for intersection)
+endpoint_closure(op, x, y, ::TrackEachEndpoint) = op(x,y)
+# alaternatively, just note the static type used to determine all endpoints
+endpoint_closure(op, x, y, tracking::TrackStatically) = tracking
+# track_closure: indicates if a boolean flag is used to track closures
+track_closures(::TrackEachEndpoint) = true
+track_closures(::TrackStatically) = false
+
+# simple endpoints are used for `TrackStatically`, when we don't want to keep
+# track of the open/closed nature of intervals. Without SimpleEndpoint the
+# return value for mergetypes would always be an array of abstract types; this
+# allows the result to be type stable when possible
+struct SimpleEndpoint{T}
+    left::Bool
+    val::T
+end
+endpoint(x::SimpleEndpoint) = x.val
+isleft(x::SimpleEndpoint) = x.left
+Base.isless(x::SimpleEndpoint, y::SimpleEndpoint) = isless(x.val, y.val)
+Base.:(==)(x::SimpleEndpoint, y::SimpleEndpoint) = x.val == y.val
+
+endpoint_types(::TrackEachEndpoint) = Endpoint
+endpoint_types(::TrackStatically{T}) where T = SimpleEndpoint{T} # see just below
+
+# `unbunch/bunch`: the generic operation used to implement all set operations operates on a
+# series of sorted endpoints (see `mergesets` below); this first requires that
+# all vectors of sets be represented by their endpoints. The functions unbunch
+# and bunch convert between an interval and an endpoint representation
+
+# when tracking endpoints dynamically we return a different type for left and right
+function unbunch_helper(interval, enumerate, i, ::TrackEachEndpoint)
+    if enumerate
+        return [(i, LeftEndpoint(interval)), (i, RightEndpoint(interval))]
+    else
+        return [LeftEndpoint(interval), RightEndpoint(interval)]
+    end
+end
+
+# when statically tracking endpoints we ignore the closed/open nature of each
+# endpoint using `SimpleEndpoint`
+function unbunch_helper(interval, enumerate, i, ::TrackStatically)
+    if enumerate
+        return [(i, SimpleEndpoint(true, first(interval))), (i, SimpleEndpoint(false, last(interval)))]
+    else
+        return [SimpleEndpoint(true, first(interval)), SimpleEndpoint(false, last(interval))]
+    end
+end
+
+function unbunch(interval::AbstractInterval, tracking::EndpointTracking; enumerate=false, lt=isless) where T
+    return unbunch_helper(interval, enumerate, 1, tracking)
+end
+function unbunch(intervals::AbstractIntervals, tracking::EndpointTracking; enumerate=false, lt=isless)
+    filtered = filter(i -> !isempty(intervals[i]), eachindex(intervals))
+    isempty(filtered) && return Union{}[]
+    result = mapreduce(vcat, filtered) do i
+        return unbunch_helper(intervals[i], enumerate, i, tracking)
+    end
```

Comment:
An option for removing the `enumerate` argument from `unbunch_helper`
```suggestion
    enumerator = enumerate ? tuple : (i, x) -> x
    result = mapreduce(vcat, filtered) do i
        return enumerator.(i, unbunch_helper(intervals[i], tracking))
    end
```
Requires changes to `unbunch_helper` to work